mlir/lib/Dialect/Rock/Transforms/SugarToLoops.cpp

```diff
@@ -888,6 +888,42 @@ static Value computeMemRefNumElements(OpBuilder &b, Location loc,
   return result;
 }
 
+/// Atomic add for a scalar fp16. Using the CAS loop (atomicRMWOp) alternative
+/// is significantly slower so we extend the scalar in a vector and use the
+/// buffer_atomic_add_fp16 instead. We have to take care of the alignment
+/// manually
+static void atomicFp16AddAligned(OpBuilder &b, Location loc, Value data,
+                                 Value dest, ArrayRef<Value> coords) {
+  const bool useBufferOobChecks = true;
+  // Useful consts
+  Value zero = b.create<arith::ConstantIntOp>(loc, 0, 32);
+  Value one = b.create<arith::ConstantIntOp>(loc, 1, 32);
+  Value two = b.create<arith::ConstantIntOp>(loc, 2, 32);
+
+  // Extended packed data to use with the intrinsic
+  Value dataExt =
+      createZeroConstantOp(b, loc, vectorTypeOrSelf(b.getF16Type(), 2));
+  Value dataExt0 = b.create<vector::InsertElementOp>(loc, data, dataExt, zero);
+  Value dataExt1 = b.create<vector::InsertElementOp>(loc, data, dataExt, one);
+
+  // Manual alignment logic : if (addr % 2 != 0) step{AddressData}Back
+  Value stepBack = b.create<arith::SubIOp>(loc, coords[2], one);
+  Value alignment = b.create<arith::RemUIOp>(loc, coords[2], two);
```

Comment:
Also, I don't think `[2]` is guaranteed to be the coordinate you want to query on.